### PR TITLE
[SPARK-7324][SQL] Add DataFrame.dropDuplicates

### PR DESCRIPTION
**Pull request by kaka1992**
Similar to http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.drop_duplicates.html
def dropDuplicates(): DataFrame
def dropDuplicates(subset: Seq[String]): DataFrame

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala [hunk 1]
```diff
@@ -888,6 +888,21 @@ class DataFrame private[sql](
...
+   * Returns a new [[DataFrame]] without duplicates.
+   * @group dfops
+   */
+  def dropDuplicates(): DataFrame = dropDuplicates(this.columns)
+
+  /**
+   * Returns a new [[DataFrame]] without duplicates under the given columns.
+   * @group dfops
+   */
+  def dropDuplicates(subset: Seq[String]): DataFrame = {
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala]: Suppose that `distinct` is as same as `dropDuplicates` for removing duplicate rows? If they are the same, which implementation is better? `GroupedData` or `Distinct` node?

Comment (kaka1992) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala]: @viirya No, dropDuplicates is used to remove duplicate rows that are the same in some columns or in all columns (default) . The default version is as same as distinct.

Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala]: You can also select subset of columns and then do distinct?

**Comment by kaka1992** [comment]
@viirya Please test this.

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala (hunk 1 above)
Comment (kaka1992) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala]: If you do this, you can't get all columns.

**Comment by kaka1992** [comment]
@rxin Please test this.

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala [hunk 2]
```diff
@@ -888,6 +888,21 @@ class DataFrame private[sql](
...
+    import org.apache.spark.sql.functions.{first => columnFirst}
+    new GroupedData(this, subset.map(colName => resolve(colName))).agg(columns.map(columnFirst))
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala]: this is just
```scala
groupBy(subset : _*).agg(columns.map(columnFirst))
```

File: sql/core/src/main/scala/org/apache/spark/sql/GroupedData.scala [hunk 3]
```diff
@@ -161,6 +161,14 @@ class GroupedData protected[sql](df: DataFrame, groupingExprs: Seq[Expression])
     DataFrame(df.sqlContext, Aggregate(groupingExprs, aggExprs, df.logicalPlan))
   }
 
+  def agg(exprs: Seq[Column]): DataFrame = {
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/GroupedData.scala]: you don't need to add this function

**Comment by rxin** [comment]
Jenkins, test this please.

**Comment by kaka1992** [comment]
@rxin Please retest this.

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala (hunk 1 above)
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala]: I'd also

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala [hunk 4]
```diff
@@ -888,6 +888,22 @@ class DataFrame private[sql](
...
+  def dropDuplicates(): DataFrame = dropDuplicates(this.columns)
+
+  /**
+   * Returns a new [[DataFrame]] without duplicates under the given columns.
+   * @group dfops
+   */
+  def dropDuplicates(subset: Seq[String]): DataFrame = {
+    import org.apache.spark.sql.functions.{first => columnFirst}
+    val columnFirsts = columns.map(columnFirst)
+    groupBy(subset.head, subset.tail : _*).agg(columnFirsts.head, columnFirsts.tail : _*)
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala]: we should also check if subset.size == 0 or columns.size == 0, then simply return an empty data frame (there is one in SQLContext).

**Comment by adrian-wang** [comment]
I think keep the `takeFirst` parameter would make this better to understand.

**Pull request closed without merging**